File: Java/Problem_037.java
import java.util.*;

public class Main {

    static boolean[] sieve(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        Arrays.fill(isPrime, true);
        isPrime[0] = isPrime[1] = false;

        for (int p = 2; (long)p * p <= limit; p++) {
            if (isPrime[p]) {
                for (long x = (long)p * p; x <= limit; x += p) {
                    isPrime[(int)x] = false;
                }
            }
        }
        return isPrime;
    }

    static boolean isTruncatable(int p, boolean[] isPrime) {
        if (p < 10) return false;

        int x = p;
        while (x > 0) {
            if (!isPrime[x]) return false;
            x /= 10;
        }

        String s = Integer.toString(p);
        for (int i = 0; i < s.length(); i++) {
            int v = Integer.parseInt(s.substring(i));
            if (!isPrime[v]) return false;
        }

        return true;
    }

    public static void main(String[] args) {
        final int LIMIT = 1_000_000;
        boolean[] isPrime = sieve(LIMIT);

        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.add(2); q.add(3); q.add(5); q.add(7);

        ArrayList<Integer> found = new ArrayList<>();
        int[] digits = {1, 3, 7, 9};

        while (found.size() < 11) {
            int n = q.removeFirst();

            for (int d : digits) {
                int m = n * 10 + d;
                if (m <= LIMIT && isPrime[m]) {
                    q.addLast(m);
                    if (isTruncatable(m, isPrime)) found.add(m);
                }
            }
        }

        long sum = 0;
        for (int x : found) sum += x;

        System.out.println(sum); // 748317
    }
}

File: CPP/Problem_037.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
using namespace std;

static vector<char> sieve(int limit) {
    vector<char> isPrime(limit + 1, 1);
    isPrime[0] = isPrime[1] = 0;
    for (int p = 2; 1LL * p * p <= limit; ++p) {
        if (isPrime[p]) {
            for (long long x = 1LL * p * p; x <= limit; x += p) {
                isPrime[(int)x] = 0;
            }
        }
    }
    return isPrime;
}

static bool isTruncatable(int p, const vector<char>& isPrime) {
    if (p < 10) return false;

    int x = p;
    while (x > 0) {
        if (!isPrime[x]) return false;
        x /= 10;
    }

    string s = to_string(p);
    for (int i = 0; i < (int)s.size(); ++i) {
        int v = stoi(s.substr(i));
        if (!isPrime[v]) return false;
    }
    return true;
}

int main() {
    const int LIMIT = 1'000'000;
    auto isPrime = sieve(LIMIT);

    queue<int> q;
    q.push(2); q.push(3); q.push(5); q.push(7);

    vector<int> found;
    int digits[4] = {1, 3, 7, 9};

    while ((int)found.size() < 11) {
        int n = q.front(); q.pop();

        for (int d : digits) {
            int m = n * 10 + d;
            if (m <= LIMIT && isPrime[m]) {
                q.push(m);
                if (isTruncatable(m, isPrime)) found.push_back(m);
            }
        }
    }

    long long sum = 0;
    for (int x : found) sum += x;

    cout << sum << "\n"; // 748317
    return 0;
}
